RefactoringsExperiments/src/main/java/ru/hse/kirilenko/refactorings/code_models/Fragment.java

```diff
@@ -0,0 +1,422 @@
+            return (StringUtils.countMatches(getMethod(), '\n') + 1);
+        }
+
+        /**
+         * Returns body of the fragment (sequence of statements)
+         * w/o leading and trailing curly braces, whitespaces, linebreaks
+         */
+        private String getBody() {
+            return clearCode(this.methodDeclaration.getBody().toString());
+        }
+
+        private String getMethod() {
+            return this.methodDeclaration.toString();
+        }
+
+        private void writeFeatures(FileWriter fw) throws IOException {
+            int nFeatures = features.size();
+            features.sort(Comparator.comparingInt(ICSVItem::getId));
+            for (int i = 0; i < nFeatures; ++i) {
+                fw.append(String.valueOf(features.get(i).getValue()));
+                fw.append(';');
+            }
+            fw.append(String.valueOf(this.score)); //Score with 4 decimal places
+            fw.append(";");
+            fw.append("-\n");
+        }
+
+        private void process(FileWriter fw) throws IOException {
+            keywordFeaturesComputation();
+            historicalFeaturesComputation();
+            couplingFeaturesComputation();
+            methodDeclarationFeaturesComputation();
+            lengthFeaturesComputation();
+            rankingScoreComputation(0.1, 4);
+
+            writeFeatures(fw);
+        }
+
+    }
+
+    private boolean isValidStatement(Statement s) {
+        String linearText = clearCode(s.toString());
+        linearText = linearText.replace("\n", "").replace("\r", "");
+//        System.out.println(s.toString()+"\n-----------------------------------\n");
+        if (Pattern.matches(".*}", linearText) | Pattern.matches(".*;", linearText))
+            return true;
+        else
+            return false;
+    }
+
+    private List<Statement> getSubStatements(Statement s) {
+        List<Statement> statements = new ArrayList<>();
+//        System.out.printf("Body: -----------------\n%s\n", block);
+        for (Node node : s.getChildrenNodes()) {
+            if (node instanceof Statement) {
+//                System.out.printf("Statement: ------------\n%s\n", node);
+                if (isValidStatement((Statement) node))
+                    statements.add((Statement) node);    //Here may be a problem of *chunky* statements
+            }
+//            System.out.printf("Sub-statements: ------------\n%s\n",node.getChildrenNodes());
+        }
+        return statements;
+    }
+
+    public void processFragment(int threshold, FileWriter fw) {
+        processEmbedding(threshold, methodDeclaration.getBody(), this.statements, fw);
+//        exit(0);
+    }
+
+    /**
+     * Splits fragment into sequences of statements of length
+     * greater than `threshold`, and processes each such
+     * sequence, generating a row to the specified file
+     */
+    private void processEmbedding(int threshold, Statement embedding, List<Statement> context, FileWriter fw) {
```

Comment:
why it is called "process**Embedding**"?

Reply:
Wanted to draw attention to the fact that it processes not atom-sized statements but embeddings of statements as well (probably I misuse SE definition of embedding but still)
Will change

Reply:
OK, I see. it confused me because I mean a vector representation by this [term](https://en.wikipedia.org/wiki/Word_embedding).